### PR TITLE
Use TypedDicts for logging.config.dictConfig

### LINKED ISSUE #6131: `logging.dictConfig` to accept `TypedDict` as an argument
Currently `dictConfig` accepts `Dict[str, Any]` as an argument.
https://github.com/python/typeshed/blob/3c58736d01b4befa4499e481e87bd7e8458dac20/stdlib/logging/config.pyi#L22

But, there's a clearly defined schema for its argument.
Docs: https://docs.python.org/3/library/logging.config.html#logging-config-dictschema

Maybe we can try to support it? It is really easy to mess things up in this long and nested dict.
The only downside is that `dictConfig()` might not accept some valid dicts with statically unknown shape.

### PR DESCRIPTION
**Pull request by A5rocks**
Closes #6131

Let's see the damage on mypy primer....

### CODE REVIEW

File: stdlib/logging/config.pyi [hunk 1]
```diff
@@ -19,7 +21,29 @@ DEFAULT_LOGGING_CONFIG_PORT: int
...
+    propagate: bool
+
+class _OptionalDictConfigArgs(TypedDict, total=False):
+    # these two can have custom factories (key: `()`) which can have extra keys
+    formatters: dict[str, dict[str, Any]]
+    filters: dict[str, dict[str, Any]]
+    # type checkers would warn about extra keys if this was a TypedDict
+    handlers: dict[str, dict[str, Any]]
+    loggers: dict[str, _LoggerConfiguration]
+    root: _RootLoggerConfiguration | None
```
Comment (asottile-sentry) [inline stdlib/logging/config.pyi:39]: coming to this late -- should this be `NotRequired` instead of `| None`?  running into some issues with this

Comment (srittau) [reply stdlib/logging/config.pyi:39]: Well, it is `NotRequired` by virtue of `_OptionalDictConfigArgs`  using `total=False`. (This code probably predates PEP 655 and could be updated.) The implementation uses `config.get("root")` to access this item:

https://github.com/python/cpython/blob/48dfd74a9db9d4aa9c6f23b4a67b461e5d977173/Lib/logging/config.py#L540C28-L540C28

That said, I think allowing `None` here is incidental, so if this makes problems, I'd be okay with removing it.

Comment (asottile-sentry) [reply stdlib/logging/config.pyi:39]: will send a PR!  basically having issues with `x: _DictConfigArgs` + `x["root"]["level"] = ...`

Comment (asottile-sentry) [reply stdlib/logging/config.pyi:39]: #11077 for that! we'll see how mypy-primer reacts

**Pull request merged**